src/libcore/intrinsics.rs

```diff
@@ -660,6 +660,26 @@ extern "rust-intrinsic" {
     /// Executes a breakpoint trap, for inspection by a debugger.
     pub fn breakpoint();
 
+    /// Returns `true` during constant evaluation and `false` otherwise.
+    ///
+    /// # Safety
+    ///
+    /// This intrinsic allows breaking [referential transparency] in `const fn`
+    /// and is therefore `unsafe`.
+    ///
+    /// Code that uses this intrinsic must be extremely careful to ensure that
+    /// `const fn`s remain referentially-transparent independently of when they
+    /// are evaluated.
```

Comment:
I find the reference of referential transparency here mostly unhelpful. At least, I would have no idea what exactly const code must (not) do, based on this description.

But also, this does not affect const-qualif. It doesn't actually introduce any referential intransparency. So personally I'd trim this down to "the code must do the same thing for both cases". Which leaves you wonder why this is even added.

Reply:
My understanding is that this is used in C++ to speed up the run-time code with SIMD and whatnot but that both code paths should observably behave same (for some notion of observational equivalence). https://en.cppreference.com/w/cpp/types/is_constant_evaluated

Reply:
If this SIMD code used run-time feature detection, couldn't be "just" make that report "no SIMD support" for const-eval to avoid needing any other magic intrinsic?

Either way though, this should be stated in the PR description.

Reply:
> If this SIMD code used run-time feature detection, couldn't be "just" make that report "no SIMD support" for const-eval to avoid needing any other magic intrinsic?

How would that run-time feature detection thing be implemented ?

Reply:
I thought that already existed for SIMD anyway?

Reply:
`is_x86_feature_detected!` has an `if cfg!(miri) { false }` branch in it, but it is not usable inside `const fn`s. To be able to use the macro in `const fn`s we'd need something like this, the macro uses inline assembly internally..

Reply:
> How would that run-time feature detection thing be implemented ? wink

Not sure, but I don't think we need this intrinsic for it? (And if we do, the name should be `is_miri`.)

Is the problem just that you cannot do `if` in `const fn`? We are accumulating more and more bad hacks to work around that and I'd rather put the line down *before* we add intrinsics for this reason. You can for example already do
```rust
#[cfg(miri)]
const fn is_miri() { true }
#[cfg(not(miri))]
const fn is_miri() { false }
```

Reply:
Unless I completely misunderstood `const fn`s scope and the plan is indeed to actually support inline assembly in const fns at some point, that is.

Reply:
> Unless I completely misunderstood const fns scope and the plan is indeed to actually support inline assembly in const fns at some point, that is.

😱: https://github.com/CraneStation/cranelift/issues/444#issuecomment-531574394